File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> lastSeen;

        int l = 0;
        int maxLen = 0;

        for(int r = 0; r < s.size(); r++){
            char ch = s[r];

            if(lastSeen.find(ch) != lastSeen.end() && l <= lastSeen[ch]){
                l = lastSeen[ch] + 1;
            }

            maxLen = max(maxLen, r - l + 1);
            lastSeen[s[r]] = r;
        }

        return maxLen;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        freq = {}
        left = 0
        maxLen = 0

        for right in range(len(s)):
            ch = s[right]

            if ch in freq and freq[ch] >= left:
                left = freq[ch] + 1
            
            maxLen = max(maxLen, (right - left + 1))

            freq[ch] = right

        return maxLen